File: content/snippets/typescript/interview_lcci/10/10_rank_from_stream.ts
class BinaryIndexedTree {
    private readonly tree: number[];

    constructor(size: number) {
        this.tree = new Array(size + 1).fill(0);
    }

    update(index: number, delta: number): void {
        while (index < this.tree.length) {
            this.tree[index] += delta;
            index += index & -index;
        }
    }

    query(index: number): number {
        let sum = 0;
        while (index > 0) {
            sum += this.tree[index];
            index -= index & -index;
        }
        return sum;
    }
}

class StreamRank {
    private readonly bit = new BinaryIndexedTree(50010);

    track(x: number): void {
        this.bit.update(x + 1, 1);
    }

    getRankOfNumber(x: number): number {
        return this.bit.query(x + 1);
    }
}

const streamRank = new StreamRank();
console.log(streamRank.getRankOfNumber(1));
streamRank.track(0);
console.log(streamRank.getRankOfNumber(0));

File: content/snippets/scala2/interview_lcci/10/10_rank_from_stream.scala
object Lcci1010RankFromStream {
    final class BinaryIndexedTree(size: Int) {
        private val tree = Array.fill(size + 1)(0)

        def update(index0: Int, delta: Int): Unit = {
            var index = index0
            while (index < tree.length) {
                tree(index) += delta
                index += index & -index
            }
        }

        def query(index0: Int): Int = {
            var index = index0
            var sum = 0
            while (index > 0) {
                sum += tree(index)
                index -= index & -index
            }
            sum
        }
    }

    final class StreamRank {
        private val bit = new BinaryIndexedTree(50010)

        def track(x: Int): Unit = bit.update(x + 1, 1)

        def getRankOfNumber(x: Int): Int = bit.query(x + 1)
    }

    def main(args: Array[String]): Unit = {
        val streamRank = new StreamRank
        println(streamRank.getRankOfNumber(1))
        streamRank.track(0)
        println(streamRank.getRankOfNumber(0))
    }
}
